confmap/converter/templateconverter/converter.go

```diff
@@ -0,0 +1,262 @@
+//	  my_template: |
+//	    receivers: ... # required, one or more
+//	    processors: ... # optional
+//	    pipelines: ...  # required if processors are used, otherwise optional
+//
+// Instantiate a templated receiver:
+//
+// receivers:
+//
+//	template/my_template[/name]:
+//	  parameter_one: value_one
+//	  parameter_two: value_two
+type converter struct {
+	cfg               map[string]any
+	receiverTemplates map[string]*template.Template
+}
+
+// New returns a confmap.Converter, that renders all templates and inserts them into the given confmap.Conf.
+func New() confmap.Converter {
+	return &converter{
+		receiverTemplates: make(map[string]*template.Template),
+	}
+}
+
+func (c *converter) Convert(_ context.Context, conf *confmap.Conf) error {
+	if err := c.parseTemplates(conf); err != nil {
+		return err
+	} else if len(c.receiverTemplates) == 0 {
+		return nil
+	}
+
+	c.cfg = conf.ToStringMap()
+	for templateID, parameters := range c.cfg["receivers"].(map[string]any) {
+		if !strings.HasPrefix(templateID, "template") {
+			continue
+		}
+
+		id, err := newInstanceID(templateID)
+		if err != nil {
+			return err
+		}
+
+		tmpl, ok := c.receiverTemplates[id.tmplType]
+		if !ok {
+			return fmt.Errorf("template type %q not found", id.tmplType)
+		}
+
+		cfg, err := newTemplateConfig(id, tmpl, parameters)
+		if err != nil {
+			return err
+		}
+
+		c.expandTemplate(cfg, id)
+	}
+	delete(c.cfg, "templates")
+
+	*conf = *confmap.NewFromStringMap(c.cfg)
+	return nil
+}
+
+func (c *converter) parseTemplates(conf *confmap.Conf) error {
+	if !conf.IsSet("templates") {
+		return nil
+	}
+
+	templatesMap, ok := conf.ToStringMap()["templates"].(map[string]any)
+	if !ok {
+		return fmt.Errorf("'templates' must be a map")
+	}
+	if templatesMap["receivers"] == nil {
+		return fmt.Errorf("'templates' must contain a 'receivers' section")
+	}
+
+	receiverTemplates, ok := templatesMap["receivers"].(map[string]any)
+	if !ok {
+		return fmt.Errorf("'templates::receivers' must be a map")
+	}
+
+	for templateType, templateVal := range receiverTemplates {
+		templateStr, ok := templateVal.(string)
+		if !ok {
+			return fmt.Errorf("'templates::receivers::%s' must be a string", templateType)
+		}
+		parsedTemplate, err := template.New(templateType).Parse(templateStr)
+		if err != nil {
+			return err
+		}
+		c.receiverTemplates[templateType] = parsedTemplate
+	}
+	return nil
+}
+
+func (c *converter) expandTemplate(cfg *templateConfig, id *instanceID) {
+	// Delete the reference to this template instance and
+	// replace it with the rendered receiver(s).
+	delete(c.cfg["receivers"].(map[string]any), id.withPrefix("template"))
+	for receiverID, receiverCfg := range cfg.Receivers {
+		c.cfg["receivers"].(map[string]any)[receiverID] = receiverCfg
+	}
+
+	// Special case where the template only contains receivers. In this case,
+	// we can just substitute the rendered receivers in place of the template ID.
+	if len(cfg.Processors) == 0 && len(cfg.Pipelines) == 0 {
```

Comment:
I don't have a particular design in mind for connector templates, or for using connectors in other classes of templates. My intuition is that it may make sense eventually but I think someone will have to propose a detailed set of rules for how this would work. I'd rather leave this for later unless we think there's a risk of designing ourselves into a corner by not solving it now.

---

Purely as speculation, I think would could eventually allow connectors within receiver templates by introducing a special key (maybe just `forward`) which MUST be used at least once as an exporter and MAY NOT be used as a receiver in the template. Basically, we'd relax the constraints on what can be used within a receiver template and simply require that the user indicate how data flows out of it. (Again, I'd rather leave this for later, but I'm including this to illustrate that it should be possible to extend the framework if necessary.)

```yaml
templates:
  receivers:
    my_receiver_template: |
      receivers:
        jaeger: ...
      connectors:
        spanmetrics: ...
      pipelines:
        traces:
          receivers: [ jaeger ]
          exporters: [ spanmetrics ] # current proposal does not allow exporters but we could relax the rule
        metrics:
          receivers: [ spanmetrics ]
          exporters: [ forward ] # "forward" is automatically defined and must be used at least once as an exporter
```